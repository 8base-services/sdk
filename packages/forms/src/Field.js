// @flow
import React from 'react';
import * as R from 'ramda';
import { Field as FinalField } from 'react-final-form';
import { compose, setDisplayName } from 'recompose';
import createValidate from '@8base/validator';
import type { FieldProps as FinalFieldProps } from 'react-final-form';

import { withFieldSchema } from './utils';
import type { FieldProps } from './types';

/**
 * `Field` wrapper based on `Field` from the [`react-final-form`](https://github.com/final-form/react-final-form). That accept [`FieldProps`](https://github.com/final-form/react-final-form#fieldprops) props and some extra props for easy working with 8base API.
 * @prop {FieldSchema} [fieldSchema] - The 8base API field schema.
 * @prop {string} [name] - The name of field, based on the 8base API table schema.
 */
class Field extends React.Component<FieldProps> {
  collectProps = (): FinalFieldProps => R.pipe(
    R.when(
      R.propIs(String, 'component'),
      R.dissoc ('fieldSchema'),
    ),
    R.assoc('validate', createValidate(this.props.fieldSchema)),
    R.assoc('name', this.props.name || this.props.name || this.props.fieldSchema.name),
  )(this.props);

  render() {
    const collectedProps = this.collectProps();

    // $FlowFixMe: Waiting a new version of react-final-form with typing fix (https://github.com/final-form/react-final-form/pull/284)
    return <FinalField { ...collectedProps } />;
  }
}

Field = compose(
  withFieldSchema,
  setDisplayName('Field'),
)(Field);

export { Field };
