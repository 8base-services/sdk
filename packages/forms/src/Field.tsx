// @flow
import React from 'react';
import { Field as FinalField } from 'react-final-form';
import { compose, setDisplayName } from 'recompose';
import createValidate from '@8base/validate';
import * as R from 'ramda';

import { withFieldSchema } from './utils';
import type { FieldProps } from './types';

/**
 * `Field` wrapper based on `Field` from the [`react-final-form`](https://github.com/final-form/react-final-form). That accept [`FieldProps`](https://github.com/final-form/react-final-form#fieldprops) props and some extra props for easy working with 8base API.
 * @prop {FieldSchema} [fieldSchema] - The 8base API field schema.
 * @prop {string} [name] - The name of field, based on the 8base API table schema.
 */
class Field extends React.Component<FieldProps> {
  collectProps = R.pipe(
    R.when(
      R.has('fieldSchema'),
      ({ validate, ...rest }) => ({
        name: rest.name || rest.fieldSchema.name,
        validate: typeof validate === 'function' ? (...args) => validate(...args, createValidate(rest.fieldSchema)) : createValidate(rest.fieldSchema),
        ...rest,
      }),
    ),
    R.when(
      R.propIs(String, 'component'),
      R.dissoc('fieldSchema'),
    ),
  );

  render() {
    const collectedProps = this.collectProps(this.props);

    // $FlowFixMe: Waiting a new version of react-final-form with typing fix (https://github.com/final-form/react-final-form/pull/284)
    return <FinalField { ...collectedProps } />;
  }
}

Field = compose(
  withFieldSchema,
  setDisplayName('Field'),
)(Field);

export { Field };
