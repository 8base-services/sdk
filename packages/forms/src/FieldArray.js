// @flow
import React from 'react';
import * as R from 'ramda';
import { FieldArray as FinalFieldArray } from 'react-final-form-arrays';
import { compose, setDisplayName } from 'recompose';
import type { FieldArrayProps as FinalFieldArrayProps } from 'react-final-form-arrays';

import { withFieldSchema } from './utils';
import type { FieldArrayProps } from './types';

/**
 * `FieldArray` wrapper based on `FieldArray` from the [`react-final-form-arrays`](https://github.com/final-form/react-final-form-arrays). It accepts [`FieldArrayProps`](https://github.com/final-form/react-final-form-arrays#fieldarrayprops) props.
 */
class FieldArray extends React.Component<FieldArrayProps> {
  collectProps = (): FinalFieldArrayProps => R.pipe(
    R.assoc('name', R.defaultTo(this.props.fieldSchema.name)(this.props.name)),
  )(this.props);

  render() {
    const collectedProps = this.collectProps();

    return <FinalFieldArray { ...collectedProps } />;
  }
}

FieldArray = compose(
  withFieldSchema,
  setDisplayName('FieldArray'),
)(FieldArray);

export { FieldArray };
