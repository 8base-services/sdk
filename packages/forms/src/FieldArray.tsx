import React from 'react';
import R from 'ramda';
import { FieldArray as FinalFieldArray } from 'react-final-form-arrays';
import { compose, setDisplayName } from 'recompose';
import { FieldArrayProps } from 'react-final-form-arrays';

import { withFieldSchema, WithFieldSchemaProps } from './utils';


const enhacner: any = compose(
  withFieldSchema,
  setDisplayName('FieldArray'),
);

/**
 * `FieldArray` wrapper based on `FieldArray` from the [`react-final-form-arrays`](https://github.com/final-form/react-final-form-arrays). It accepts [`FieldArrayProps`](https://github.com/final-form/react-final-form-arrays#fieldarrayprops) props.
 */
const FieldArray: React.ComponentType<FieldArrayProps> = enhacner(
  class FieldArray extends React.Component<FieldArrayProps & WithFieldSchemaProps> {
    collectProps = () => {
      const { fieldSchema, name, isEqual, ...restProps } = this.props;

      if (fieldSchema) {
        return {
          name: name || fieldSchema.name,
          fieldSchema,
          isEqual: isEqual || R.equals,
          ...restProps,
        };
      }

      return this.props;
    };

    render() {
      const collectedProps = this.collectProps();

      return <FinalFieldArray { ...collectedProps } />;
    }
  }
)

export { FieldArray };
