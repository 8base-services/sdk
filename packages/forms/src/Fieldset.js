// @flow
import React from 'react';
import { compose, setDisplayName } from 'recompose';

import { FormContext } from './FormContext';
import { withTableSchema, renderComponent } from './utils';
import type { FieldsetProps, FormContextValue } from './types';

/**
 * `Fieldset` passes relation table schema to the children fields.
 * @prop {TableSchema} [tableSchema] - The 8base API table schema.
 * @prop {string} [tableSchemaName] - The name of the 8base API table schema. Worked only if you provide schema by `SchemaContext`.
 */
class Fieldset extends React.Component<FieldsetProps> {
  collectContextValue = (): FormContextValue => {
    const { tableSchema } = this.props;

    return { tableSchema };
  };

  render() {
    const contextValue: FormContextValue = this.collectContextValue();

    const rendered = renderComponent(this.props);

    return (
      <FormContext.Provider value={ contextValue }>
        { rendered }
      </FormContext.Provider>
    );
  }
}

Fieldset = compose(
  withTableSchema,
  setDisplayName('Fieldset'),
)(Fieldset);

export { Fieldset };
