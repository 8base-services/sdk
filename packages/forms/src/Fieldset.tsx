import React from 'react';
import { compose, setDisplayName } from 'recompose';

import { FormContext } from './FormContext';
import { withTableSchema, renderComponent, WithTableSchemaProps } from './utils';
import { FieldsetProps, FormContextValue } from './types';

const enhacner: any = compose(
  withTableSchema,
  setDisplayName('Fieldset'),
);

/**
 * `Fieldset` passes relation table schema to the children fields.
 * @prop {TableSchema} [tableSchema] - The 8base API table schema.
 * @prop {string} [tableSchemaName] - The name of the 8base API table schema. Worked only if you provide schema by `SchemaContext`.
 */
const Fieldset: React.ComponentType<FieldsetProps> = enhacner(
  class Fieldset extends React.Component<FieldsetProps & WithTableSchemaProps> {
    public collectContextValue = (): FormContextValue => {
      const { tableSchema } = this.props;

      return { tableSchema };
    };

    public render() {
      const contextValue: FormContextValue = this.collectContextValue();

      const rendered = renderComponent(this.props);

      return <FormContext.Provider value={contextValue}>{rendered}</FormContext.Provider>;
    }
  },
);

export { Fieldset };
