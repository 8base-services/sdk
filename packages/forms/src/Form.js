// @flow
import React from 'react';
import * as R from 'ramda';
import { FORM_ERROR } from 'final-form';
import { Form as FinalForm } from 'react-final-form';
import arrayMutators from 'final-form-arrays';
import { compose, setDisplayName } from 'recompose';
import type { FormProps as FinalFormProps } from 'react-final-form';
import { formatDataForMutation } from '@8base/utils';
import errorCodes from '@8base/error-codes';

import { FormContext } from './FormContext';
import { withTableSchema } from './utils';
import type { FormProps, FormContextValue } from './types';

/**
 * `Form` wrapper based on `Form` from the [`react-final-form`](https://github.com/final-form/react-final-form). That accept [`FormProps`](https://github.com/final-form/react-final-form#formprops) props and some extra props for easy working with 8base API.
 * @prop {TableSchema} [tableSchema] - The 8base API table schema.
 * @prop {string} [tableSchemaName] - The name of the 8base API table schema.
 */
class Form extends React.Component<FormProps> {
  static defaultProps = {
    mutators: {},
  };

  collectProps = (): FinalFormProps => {
    const { mutators, tableSchema, type, schema, onSubmit, ...restProps } = this.props;

    const collectedProps = {
      mutators: R.merge(arrayMutators, mutators),
      tableSchema,
      onSubmit,
      ...restProps,
    };

    if (type && tableSchema && schema) {
      collectedProps.onSubmit = async (data, ...rest) => {
        try {
          await onSubmit(formatDataForMutation(type, tableSchema.name, data, schema), ...rest);
        } catch (e) {
          if (e.graphQLErrors) {
            const error = e.graphQLErrors[0];

            if (error.code === errorCodes.ValidationErrorCode) {
              return error.details;
            }

            return { [FORM_ERROR]: error.message };
          }
        }
      };
    }

    return collectedProps;
  };

  collectContextValue = (): FormContextValue => {
    const { tableSchema } = this.props;

    return { tableSchema };
  };

  render() {
    const props: FinalFormProps = this.collectProps();
    const contextValue: FormContextValue = this.collectContextValue();

    return (
      <FormContext.Provider value={ contextValue }>
        <FinalForm { ...props } />
      </FormContext.Provider>
    );
  }
}

Form = compose(
  withTableSchema,
  setDisplayName('Form'),
)(Form);

export { Form };
