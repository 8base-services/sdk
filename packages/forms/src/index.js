// @flow
export { Form } from './Form';
export { Field } from './Field';
export { Fieldset } from './Fieldset';
export { FieldArray } from './FieldArray';
export { SchemaContext } from './SchemaContext';
export { FormContext } from './FormContext';
