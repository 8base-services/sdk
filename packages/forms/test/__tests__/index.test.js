import React from 'react';
import TestUtils from 'react-dom/test-utils';
import renderer from 'react-test-renderer';

import { Form, Field, FieldArray, SchemaContext } from '../../src';

it('As a developer, I can prefill implemented from with default values from table schema.', () => {
  const tableSchema = {
    name: 'tableSchema',
    displayName: 'tableSchema',
    isSystem: false,
    fields: [{
      name: 'field',
      displayName: 'Text Required Unformatted Field',
      description: null,
      fieldType: 'TEXT',
      fieldTypeAttributes: {
        format: 'UNFORMATTED',
        fieldSize: 100,
      },
      isList: false,
      isRequired: true,
      isUnique: false,
      defaultValue: 'Default Value',
      relation: null,
    }],
  };

  const TestFieldComponent = jest.fn(() => null);

  const form = renderer.create(
    <Form onSubmit={ jest.fn() } tableSchema={ tableSchema } prefillInitialValues>
      {
        ({ handleSubmit }) => (
          <form onSubmit={ handleSubmit }>
            <Field fieldSchemaName={ tableSchema.fields[0].name } component={ TestFieldComponent } />
          </form>
        )
      }
    </Form>,
  );

  expect(TestFieldComponent).toHaveBeenCalledTimes(1);
  expect(TestFieldComponent.mock.calls[0][0].input.value).toEqual('Default Value');
});

it('As a developer, I can get access to schema of the each field in the field component.', () => {
  const tableSchema = {
    name: 'tableSchema',
    displayName: 'tableSchema',
    isSystem: false,
    fields: [{
      name: 'field',
      displayName: 'Text Required Unformatted Field',
      description: null,
      fieldType: 'TEXT',
      fieldTypeAttributes: {
        format: 'UNFORMATTED',
        fieldSize: 100,
      },
      isList: false,
      isRequired: false,
      isUnique: false,
      defaultValue: 'Default Value',
      relation: null,
    }],
  };

  const TestFieldComponent = jest.fn(() => null);

  const form = renderer.create(
    <Form onSubmit={ jest.fn() } tableSchema={ tableSchema }>
      {
        ({ handleSubmit }) => (
          <form onSubmit={ handleSubmit }>
            <Field fieldSchemaName={ tableSchema.fields[0].name } component={ TestFieldComponent } />
          </form>
        )
      }
    </Form>,
  );

  expect(TestFieldComponent).toHaveBeenCalledTimes(1);
  expect(TestFieldComponent.mock.calls[0][0].fieldSchema).toEqual(tableSchema.fields[0]);
});

it('As a developer, I can get access to table schema from the render props.', () => {
  const tableSchema = {
    name: 'tableSchema',
    displayName: 'tableSchema',
    isSystem: false,
    fields: [{
      name: 'field',
      displayName: 'Text Required Unformatted Field',
      description: null,
      fieldType: 'TEXT',
      fieldTypeAttributes: {
        format: 'UNFORMATTED',
        fieldSize: 100,
      },
      isList: false,
      isRequired: false,
      isUnique: false,
      defaultValue: 'Default Value',
      relation: null,
    }],
  };

  const TestFieldComponent = jest.fn(() => null);

  const renderProp = jest.fn(({ handleSubmit }) => (
    <form onSubmit={ handleSubmit }>
      <Field fieldSchemaName={ tableSchema.fields[0].name } component={ TestFieldComponent } />
    </form>
  ));

  const form = renderer.create(
    <Form onSubmit={ jest.fn() } tableSchema={ tableSchema }>
      { renderProp }
    </Form>,
  );

  expect(renderProp).toHaveBeenCalledTimes(1);
  expect(renderProp.mock.calls[0][0].tableSchema).toEqual(tableSchema);
});

it('As a developer, I can get access to validation errors from field components.', () => {
  const TestFieldComponent = jest.fn(() => null);

  const onSubmit = jest.fn();

  const tableSchema = {
    name: 'tableSchema',
    displayName: 'tableSchema',
    isSystem: false,
    fields: [{
      name: 'field',
      displayName: 'Text Required Unformatted Field',
      description: null,
      fieldType: 'TEXT',
      fieldTypeAttributes: {
        format: 'UNFORMATTED',
        fieldSize: 100,
      },
      isList: false,
      isRequired: true,
      isUnique: false,
      defaultValue: null,
      relation: null,
    }, {
      name: 'fieldArray',
      displayName: 'List of Text Required Unformatted Fields',
      description: null,
      fieldType: 'TEXT',
      fieldTypeAttributes: {
        format: 'UNFORMATTED',
        fieldSize: 2,
      },
      isList: true,
      isRequired: true,
      isUnique: false,
      defaultValue: null,
      relation: null,
    }],
  };

  const form = TestUtils.renderIntoDocument(
    <Form onSubmit={ onSubmit } tableSchema={ tableSchema } initialValues={{ fieldArray: ['Award'] }}>
      {
        ({ handleSubmit }) => (
          <form onSubmit={ handleSubmit }>
            <Field fieldSchemaName={ tableSchema.fields[0].name } component={ TestFieldComponent } />
            <FieldArray fieldSchemaName={ tableSchema.fields[1].name }>
              {
                ({ fields }) => (
                  fields.map((name) => (
                    <Field key={ name } fieldSchemaName={ tableSchema.fields[1].name } name={ name } component={ TestFieldComponent } />
                  ))
                )
              }
            </FieldArray>
            <button type="onSubmit" />
          </form>
        )
      }
    </Form>,
  );

  submitForm(form);

  expect(onSubmit).toHaveBeenCalledTimes(0);

  expect(TestFieldComponent).toHaveBeenCalledTimes(6);
  expect(TestFieldComponent.mock.calls[4][0].meta.error).toEqual('Value is required');
  expect(TestFieldComponent.mock.calls[5][0].meta.error).toEqual('Maximum allowed field size is 2. It was exceeded.');
});

it('As a developer, I can pass table schema to multiple forms via schema context provider.', () => {
  const schema = [{
    name: 'tableSchema',
    displayName: 'tableSchema',
    isSystem: false,
    fields: [{
      name: 'field',
      displayName: 'Text Required Unformatted Field',
      description: null,
      fieldType: 'TEXT',
      fieldTypeAttributes: {
        format: 'UNFORMATTED',
        fieldSize: 100,
      },
      isList: false,
      isRequired: true,
      isUnique: false,
      defaultValue: null,
      relation: null,
    }],
  }];

  const form = renderer.create(
    <SchemaContext.Provider value={{ schema }}>
      <Form onSubmit={ jest.fn() } tableSchemaName="tableSchema" initialValues={{ field: 'caupones' }}>
        {
          ({ handleSubmit }) => (
            <form onSubmit={ handleSubmit }>
              <Field fieldSchemaName={ schema[0].fields[0].name } component="input" />
            </form>
          )
        }
      </Form>
      <Form onSubmit={ jest.fn() } tableSchemaName="tableSchema" initialValues={{ field: 'tubiporous' }}>
        {
          ({ handleSubmit }) => (
            <form onSubmit={ handleSubmit }>
              <Field fieldSchemaName={ schema[0].fields[0].name } component="input" />
            </form>
          )
        }
      </Form>
    </SchemaContext.Provider>,
  );

  expect(form.toJSON()).toMatchSnapshot();
});
