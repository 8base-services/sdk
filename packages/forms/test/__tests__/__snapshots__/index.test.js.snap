// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`As a developer, I can pass table schema to multiple forms via schema context provider. 1`] = `
Array [
  <form
    onSubmit={[Function]}
  >
    <input
      name="field"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      value="caupones"
    />
  </form>,
  <form
    onSubmit={[Function]}
  >
    <input
      name="field"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      value="tubiporous"
    />
  </form>,
]
`;
